logReader: manually commit offset when using kafka logConsumer

welansari opened the pull request:
Kafka log consumer currently auto-commits messages after they are consumed. This is an issue as it means we loose messages after restart or crash even if their processing didn't finish.

Like we do with other logConsumer implementations, we now manually store offsets after the end of each batch. The auto-commit mechanism commits offsets stored locally after each 5 seconds.

Contrary to the BackbeatConsumer, there is no risk of offsets being committed in disorder as only one batch is processed at a time, and a partition is only assigned to a single instance of the QueuePopulator.

Issue: BB-698

Commit:
logReader: manually commit offset when using kafka logConsumer

Kafka log consumer currently auto-commits messages after they
are consumed. This is an issue as it means we loose messages
after restart of crash even if their processing didn't finish.

Like we do with other logConsumer implementations, we now manually
store offsets after the end of each batch. The auto-commit mechanism
commits offsets stored locally each interval.

Contrary to the BackbeatConsumer, there is no risk of offsets being
committed in disorder as only one batch is processed at a time, and
a partition is only assigned to a single instance of the QueuePopulator.

Issue: BB-698

bert-e commented:
# Hello kerkesni,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/backbeat/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-698 contains:


* `9.0.13`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `9.0.13`

* `9.1.0`


Please check the `Fix Version/s` of BB-698, or the target
branch of this pull request.

bert-e commented:
# Request integration branches



Waiting for integration branch creation to be requested by the user.

To request integration branches, please comment on this pull request with the following command:

```
/create_integration_branches
```

Alternatively, the `/approve` and `/create_pull_requests` commands will automatically
create the integration branches.

bert-e commented:
# Request integration branches



Waiting for integration branch creation to be requested by the user.

To request integration branches, please comment on this pull request with the following command:

```
/create_integration_branches
```

Alternatively, the `/approve` and `/create_pull_requests` commands will automatically
create the integration branches.

francoisferrand commented on lib/queuePopulator/KafkaLogConsumer/LogConsumer.js:
```diff
@@ -20,17 +20,19 @@ class LogConsumer {
     /**
      * Get partition offsets
-     * @returns {string} stored partition offsets
+     * Offsets are stored in kafka are not managed
```
typo?
```suggestion
     * Offsets are stored in kafka and not managed
```

francoisferrand commented on lib/queuePopulator/KafkaLogConsumer/LogConsumer.js:
```diff
@@ -42,13 +44,17 @@ class LogConsumer {
+            // after the batch processing is fully completed.
+            'enable.auto.offset.store': false,
+            // Default auto-commit interval is 5 seconds
+            'enable.auto.commit': true,
```
with `auto-commit`, we will still keep a window (5 seconds) of uncertainty:
- is it the same with backbeatconsumer? I don't see the flag there...
- in backbeatConsume, I think we have code to ensure we "flush" the commited offset when needed : should we do the same here?

welansari replied:
> is it the same with backbeatconsumer? I don't see the flag there...

Yes, we have the same in `BackbeatConsumer`, the option is enabled by default i just decided to included it here for clarity.

> in backbeatConsume, I think we have code to ensure we "flush" the commited offset when needed : should we do the same here?

Yes, i should handle rebalance to avoid replay cases

francoisferrand commented on lib/queuePopulator/KafkaLogConsumer/LogConsumer.js:
```diff
@@ -42,13 +44,17 @@ class LogConsumer {
         const consumerParams = {
-            'enable.auto.offset.store': true,
+            // Manually manage storing offsets to ensure they are only stored
+            // after the batch processing is fully completed.
```
bonus issue: this will raise the issues we had (and fixed) in BackbeatConsumer around handling of rebalance, i.e. the need to delay the rebalance until the batch has been completed and offsets commited accordingly, and the need to detect/handle "slow tasks" which could block all further processing... It may be done in another PR (or maybe we could reuse BackbeatConsumer for KafkaLogConsumer?), but we probably can't go to prod without that kind of things...

francoisferrand commented on lib/queuePopulator/KafkaLogConsumer/LogConsumer.js:
```diff
@@ -155,16 +180,39 @@ class LogConsumer {
+            // ending and returning the stream
+            this._listRecordStream.end();
+            return cb(null, { log: this._listRecordStream, tailable: false });
+        });
```
nit: i think the original flow was better, as it managed the callback (`cb`) in a single place, with no added condition (`if`) to handle error... or was there an issue? (we can also do `res?.[3]` if res could be null)

francoisferrand commented on lib/queuePopulator/LogReader.js:
```diff
@@ -712,25 +712,50 @@ class LogReader {
         }
+
+        // Handle offset managed externally (e.g., Kafka)
+        if (!this.isOffsetManaged() && this.logConsumer.storeOffsets) {
```
When `!this.isOffsetManaged()`, then `storeOffsets` should always be defined, right?
If this is the case, then this option is redundant (an assertion in constructor is enough)

This would allow a much nicer flow:
```
    if (!this.isOffsetManaged()) {
        this.logConsumer.storeOffsets();
    } else if (this._shouldUpdateLogOffset(logRes, nextLogOffset)) {
        ...
    }
```

welansari replied:
> When !this.isOffsetManaged(), then storeOffsets should always be defined, right?

Currently yes, the only logReader for which we don't manage offsets is the Kafka logReader.

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-698 contains:


* `9.0.14`

* `9.1.0`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `9.0.15`

* `9.1.0`


Please check the `Fix Version/s` of BB-698, or the target
branch of this pull request.

Commit:
handle rebalance in the Kafka logConsumer

Issue: BB-698

Commit:
fail healthcheck when Kafka logConsumer is disconnected

Issue: BB-698

Commit:
wait for offsets to be committed before shutting down the QueuePopulator

Issue: BB-698

Commit:
add unit test

Issue: BB-698

bert-e commented:
# Request integration branches



Waiting for integration branch creation to be requested by the user.

To request integration branches, please comment on this pull request with the following command:

```
/create_integration_branches
```

Alternatively, the `/approve` and `/create_pull_requests` commands will automatically
create the integration branches.

francoisferrand commented on lib/queuePopulator/KafkaLogConsumer/LogConsumer.js:
```diff
@@ -136,8 +219,20 @@ class LogConsumer {
+                this._listRecordStream.write(message);
+            });
+            // format offsets for commit
+            this._topicPartition = Array.from(topicPartition.values());
```
Should we not `merge` this map with the existing `this._topicPartition` ?
e.g. in `_consumeKafkaMessages` is called multiple times before `storeOffsets()` is called

(maybe this does not happen, but it is not clear from the prototype of LogConsumer that it MUST work this way...)

In that case, may be better to keep a map in LogConsumer, and build the array only as a temporary varialbe in storeOffsets

welansari replied:
The queuePopulator doesn't support parallel processing of batches, only a single batch of messages is consumed, processed and then committed at a time. I don't see a need to support that currently.

I enforced this by not allowing consumption of new messages when a batch is in progress.

francoisferrand commented on lib/queuePopulator/KafkaLogConsumer/LogConsumer.js:
```diff
@@ -42,70 +46,141 @@ class LogConsumer {
-            if (err) {
-                this._log.error('Error while getting offsets', {
-                    method: 'LogConsumer._storeCurrentOffsets',
+    _onRebalance(err, assignment) {
```
should this update metrics: `KafkaBacklogMetrics.onRebalance(this._topic, this._groupId, status)` ?
(or is that metrics specific to BackbeatConsumer)

welansari replied:
No the metric is not specific to BackbeatConsumer, i added it

francoisferrand commented on lib/queuePopulator/KafkaLogConsumer/LogConsumer.js:
```diff
@@ -42,70 +46,141 @@ class LogConsumer {
+                    return;
+                }
+                // wait for consumed messages to finish being processed
+                if (this._topicPartition && this._topicPartition.length > 0) {
```
instead of polling, can't we trigger this callback when the topic gets empty?
similar to BackbeatConsumer, we could create a `drain` callback which is called when `_topicPartition` is reset ; and set it when needed in rebalance.

francoisferrand commented on lib/queuePopulator/KafkaLogConsumer/LogConsumer.js:
```diff
@@ -153,18 +248,80 @@ class LogConsumer {
+                if (this._consumer?.isConnected()) {
+                    this._consumer.disconnect();
+                    this._consumer.once('disconnected', () => next());
+                    return null;
```
should we not do it the other way, to avoid race conditions?
(node is thread safe, so probably fine as is... but could change with futurechanges...)
```suggestion                  
                    this._consumer.once('disconnected', () => next());
                    this._consumer.disconnect();
                    return null;
```

(unless the semantics is that setting `once('disconnected')` callback will immediately invoke the callback if the consumer is alreadyd disconnected)

francoisferrand commented on lib/queuePopulator/LogReader.js:
```diff
@@ -111,10 +111,18 @@ class LogReader {
-            Object.values(this._producers),
-            (producer, done) => producer.close(done),
-            cb);
+        async.parallel([
```
should we not first (in series!) stop the consumer -waiting for all 'current' messages to have been consumer- then stop the producers?

francoisferrand commented on tests/unit/lib/queuePopulator/LogReader.spec.js:
```diff
@@ -599,30 +599,34 @@ describe('LogReader', () => {
                 currentOffset: 0,
                 nextLogOffset: 0,
                 shouldSave: false,
             },
```
should we really remove that test:
- that would mean we changed the behavior, which seems weird?
- the other tests seem to have similar conditions ("offsets not managed"), so this not a corner case we dropped
- if we remove it from here, we should probably create a new one (or multiple, if offset are/aren't managed)

welansari replied:
The behaviour did change for the non managed logConsumer (Kafka) and it's not weird, the whole point of this PR is to now handle how/when the offsets get committed. Before, we used to do nothing and the offsets would be committed as soon as messages were consumed, now we do it manually after each batch of messages finished processing.

Commit:
fix signals not being propagated

the `-g` option allows propagating the signal
to the entire child process group instead of
only the direct child which might not handle
signal propagation, which is the case for the
shell created by npm.

Issue: BB-698

Commit:
force sequential consumption in the Kafka logConsumer

Ensuring a single batch of messages is consumed, processed
and committed a time in the Kafka logConsumer. This is the
current behaviour of the QueuePopulator, as it does not
support parallel consumption and processing of batches.

Issue: BB-698

Commit:
refactor rebalance handler of the Kafka logConsumer

Issue: BB-698

Commit:
set listeners before disconnect and unsubscribe to avoid race conditions

Issue: BB-698

welansari commented:
/create_integration_branches

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/9.1/bugfix/BB-698` with contents from `bugfix/BB-698`
and `development/9.1`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/9.1/bugfix/BB-698 origin/development/9.1
 $ git merge origin/bugfix/BB-698
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/9.1/bugfix/BB-698
```





*The following options are set:* **create_integration_branches**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **create_integration_branches**

welansari commented:
/approve

bert-e commented:
# Build failed



The [build](https://github.com/scality/backbeat/actions?query=branch%3Aw/9.1/bugfix/BB-698) for [commit](https://github.com/scality/backbeat/commit/b80c527f1698891153ecec6ca1a488a9487f47a2) did not succeed in branch [w/9.1/bugfix/BB-698](https://github.com/scality/backbeat/compare/development/9.1...w/9.1/bugfix/BB-698)



*The following options are set:* **approve, create_integration_branches**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/9.0`

* :heavy_check_mark: `development/9.1`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.6`



Please check the status of the associated issue BB-698.

Goodbye kerkesni.




*The following options are set:* **approve, create_integration_branches**
